crates/kitsune_p2p/kitsune_p2p/tests/tests.rs

```diff
@@ -2,88 +2,1148 @@ mod common;
+        .update_tuning_params(|mut c| {
+            // 3 seconds between gossip rounds, to keep the test fast
+            c.gossip_peer_on_success_next_gossip_delay_ms = 1000 * 3;
+            c
+        });
+
+    let sender_b = harness_b
+        .spawn()
+        .await
+        .expect("should be able to spawn node");
+
+    let agent_a = harness_a.create_agent().await;
+
+    sender_a
+        .join(space.clone(), agent_a.clone(), None, None)
+        .await
+        .unwrap();
+
+    let agent_b = harness_b.create_agent().await;
+
+    sender_b
+        .join(space.clone(), agent_b.clone(), None, None)
+        .await
+        .unwrap();
+
+    // Wait for the nodes to discover each other before publishing
+    wait_for_connected(sender_a.clone(), agent_b.clone(), space.clone()).await;
+    wait_for_connected(sender_b.clone(), agent_a.clone(), space.clone()).await;
+
+    sender_a
+        .broadcast(
+            space.clone(),
+            basis,
+            KitsuneTimeout::from_millis(5_000),
+            BroadcastData::Publish {
+                source: agent_a.clone(),
+                op_hash_list: harness_a
+                    .op_store()
+                    .read()
+                    .iter()
+                    .map(|o| o.clone().into())
+                    .collect(),
+                context: FetchContext::default(),
+            },
+        )
+        .await
+        .unwrap();
+
+    tokio::time::timeout(std::time::Duration::from_secs(600), {
+        let op_store_b = harness_b.op_store();
+        async move {
+            loop {
+                if op_store_b.read().len() >= num_ops {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+            }
+        }
+    })
+    .await
+    .expect("Expected B to get all ops but this hasn't happened");
+
+    assert_eq!(num_ops, harness_b.op_store().read().len());
+
+    let events = harness_b
+        .drain_legacy_host_events()
+        .await
+        .into_iter()
+        .filter_map(|e| match e {
+            RecordedKitsuneP2pEvent::ReceiveOps { ops, .. } => Some(ops),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    // Expect at least one event per op
+    assert!(events.len() >= num_ops);
+
+    // The total of receieved ops must be the same as the total published. This prevents the test from quietly receiving duplicates.
+    assert_eq!(num_ops, harness_b.op_store().read().len());
+
+    // TODO Can't use this assertion, duplicate ops are usually sent during this test.
+    //      The `incoming_dht_ops_workflow` is what would deal with this problem in the Holochain host implementation but it'd be a nice guarantee if
+    //      Kitsune didn't hand the host ops that it already has. It's not a lot of overhead to check later but the ghost actor queues are a limited
+    //      resource.
+    // assert_eq!(0, harness_b.duplicate_ops_received_count());
```

Comment:
This I'm a bit unsure about. It's always going to be the case that some duplicate information is received one way or another. So the host is going to have to check, or we are going to have to ask the host to check before handing it over. To me, it probably makes the most sense for the host to be resilient to this. But in such a simple test it feels like this shouldn't happen and it might be worth digging in and looking at why it does?

Reply:
For context, I originally used a `Vec` in the host to collect received ops. That resulted in failed test assertions because of the duplicates. I am now checking before inserting to avoid duplicates and keeping track of the number of times that happened instead.